File: Modules/SMES/MES/SIE.Web.MES/DataBarcode/Commands/DataBarcodeLabelPrintCommand.js
SIE.defineCommand('SIE.Web.MES.DataBarcode.Commands.DataBarcodeLabelPrintCommand', {
    meta: { text: "数据条码化打印", group: "edit", iconCls: "icon-PrintData icon-blue" },
    canExecute: function (view) {
        var selectModels = view.getSelection();
        if (selectModels.length == 0) return false;
        return true;
    },
    execute: function (view, source) {
        var huId = view.getCurrent().data.Id;
        SIE.AutoUI.getMeta({
            model: "SIE.Web.MES.DataBarcode.ViewModels.DataBarcodeViewModels",
            ignoreCommands: false,
            isDetail: true,
            ignoreQuery: true,
            viewGroup: "DetailsView",
            callback: function (res) {
                var mainBolck;
                if (res.mainBolck)
                    mainBolck = res.mainBolck;
                else
                    mainBolck = res;
                var detailView = SIE.AutoUI.createDetailView(mainBolck);
                detailView.token = view.getToken();
                var entity = new detailView._model();
                //entity.setIsSubBatch(false);
                var ui = detailView.getControl();
                SIE.invokeDataQuery({
                    type: "SIE.Web.MES.DataBarcode.DataQuerys.DataBarcodeDataQuery",
                    method: "GetReprintInfo",
                    params: [huId],
                    async: false,
                    token: view.token,
                    callback: function (res) {
                        if (res.Success) {
                            var defaultInfo = res.Result;
                            //entity.setTemplateId(defaultInfo.data.items[0].data.TemplateId);
                            //entity.setTemplateId_Display(defaultInfo.data.items[0].data.TemplateId_Display);
                            detailView.setData(entity);
                            var win = SIE.Window.show({
                                title: "数据条码化打印".t(),
                                width: 480,
                                height: 250,
                                items: ui,
                                id: "ReprintBatch001",
                                callback: function (btn) {
                                    if (btn == "确定".t()) {
                                        var reprintInfo = detailView.getData().data;
                                        if (!reprintInfo.TemplateId || reprintInfo.TemplateId <= 0) {
                                            SIE.Msg.showMessage("打印模板不能为空".L10N());
                                            return false;
                                        }
                                        view.execute({
                                            data: { DataBarcodeIds: view.getSelectionIds(), TemplateId: reprintInfo.TemplateId },
                                            success: function (res) {
                                                win.close();

                                                var rstPrint = res.Result;

                                                if (rstPrint.ErrMsg !== '') {
                                                    SIE.Msg.showError(rstPrint.ErrMsg);
                                                    return false;
                                                } else {
                                                    win.close();

                                                    var printCmpt = new SIE.Web.Common.Prints.Report.WebReportComponents({
                                                        ReportType: rstPrint.Type,
                                                        ReportData: {
                                                            path: rstPrint.Url, content: rstPrint.Url
                                                        }
                                                    });

                                                    var cfg = printCmpt.getExtTarget();

                                                    if (cfg && cfg.printCallback) {
                                                        cfg.printCallback(printCmpt);
                                                    }
                                                    else {
                                                        var param = printCmpt.getPrintParams();
                                                        if (!printCmpt.hasError()) {
                                                            var printUrl = printCmpt.getPrintUrl();
                                                            if (!printCmpt.hasError())
                                                                CRT.Workbench.showPageDialog({ id: 'ReprintBatch_rpt', text: "物料标签打印".t(), method: 'POST', url: printUrl, params: param });
                                                        }
                                                    }

                                                    view.getSelectionModel().deselectAll();

                                                    view._parent.reloadData();
                                                }
                                            },
                                            error: function (res) {
                                                SIE.Msg.showMessage(res.Message);
                                                return false;
                                            }
                                        });
                                    }
                                }
                            });
                        }
                    }
                });
            }
        });
    }
});

File: Modules/SMES/MES/SIE.Web.MES/DataBarcode/Commands/DataBarcodeLabelPrintCommand.cs

using SIE.Barcodes.Barcodes.ViewModels;
using SIE.Barcodes.Printables;
using SIE.Common.Prints;
using SIE.Domain;
using SIE.MES.DataBarcode;
using SIE.Packages.ItemLabels;
using SIE.Web.Command;
using SIE.Web.Common.Prints;
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;

namespace SIE.Web.MES.DataBarcode.Commands
{
        /// <summary>
        /// 批次信息补打
        /// </summary>
        public class DataBarcodeLabelPrintCommand : ListViewCommand
        {
            /// <summary>
            /// 补打命令执行方法
            /// </summary>
            /// <param name="args">args</param>
            /// <param name="scope">scope</param>
            /// <returns>补打结果</returns>
            protected override object Excute(ViewArgs args, string scope)
            {
                var rstPrint = new RstBarcodePrint();
                rstPrint.ErrMsg = string.Empty;

                try
                {
                    var entity = args.Data.ToJsonObject<BatchReprintDataModel>();
                    var template = RF.GetById<PrintTemplate>(entity.TemplateId);
                    Check.NotNull(template, nameof(PrintTemplate));

                //if (template.EntityType != typeof(WipBatchPrintable).GetQualifiedName())
                //{
                //    throw new ValidationException("打印模板的类型必须是生产批次".L10N());
                //}

                rstPrint.Type = template.Type;

                var wipBatchs = RT.Service.Resolve<DataBarcodeController>().GetDataBarcodePrintDatas(entity.DataBarcodeIds);
                if (wipBatchs[0].BarcodeType != template.FileName.Replace(".siedev", "")) {
                    throw new ValidationException("打印模板的类型【"+ wipBatchs[0].BarcodeType+ "】与选择的类型【"+ template.FileName.Replace(".siedev", "") + "】不一致".L10N());
                }
                var report = ReportFactory.Current.GetReportByExtension(template.Type);
                
                    var printable = new DataBarcodePrintable();

                    rstPrint.Url = report.PrintProcess(printable, template.Id, template.Content, () =>
                    {
                        return wipBatchs;
                    });
                }
                catch (Exception exc)
                {
                    rstPrint.ErrMsg = exc.Message;
                }

                return rstPrint;
            }


            /// <summary>
            /// 批次补打数据类
            /// </summary>
            public class BatchReprintDataModel
            {
                /// <summary>
                /// 选中条码Id列表
                /// </summary>
                public List<double> DataBarcodeIds { get; set; }

                /// <summary>
                /// 补打模板Id
                /// </summary>
                public double TemplateId { get; set; }
            }
        }
    }
